pkg/rbd/controllerserver.go

```diff
@@ -136,6 +145,11 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		// request
 		if exVol.VolSize >= req.GetCapacityRange().GetRequiredBytes() {
 			// existing volume is compatible with new request and should be reused.
+
+			if err = storeVolumeMetadata(exVol, cs.MetadataStore); err != nil {
```

Comment:
@gman0 I have a question here, as this` exVol` is retrieved from the metadataStore why we need to store it back?

Reply:
Well yes and no... `rbdVolumes` (and snapshots) is initialized from the metadata cache - but after that, it's populated by the `CreateVolume` function itself https://github.com/gman0/ceph-csi/blob/d9f71d3887f4eaed3cfa34de8bbd1686163d0a32/pkg/rbd/controllerserver.go#L178

Reply:
But my question is why we have store it back. I think we can update the local cache once we store the Metadata on configmap is successful.
 By doing this we can avoid extra step of storing if volume is already present, isn't it?

Reply:
Oops, you're right! For whatever reason I've completely missed this bit https://github.com/gman0/ceph-csi/blob/d9f71d3887f4eaed3cfa34de8bbd1686163d0a32/pkg/rbd/controllerserver.go#L172-L176

which checks for the existence of the image too...nice catch!

Reply:
@gman0 even with the current design, suppose CSI pod restarts after creating volume in the backend and before storing metadata on config map, we may end up in having stale volumes in the backend. do we have a CLI command  in cephfs to list the volumes (by checking at  the  backend we can assure that there won't be any stale volumes even if  CSI restarts) (note: PVC create/delete may  become a slow  process)

Reply:
reading more through the code, the only issue we are having with restart (as explained above) is volumeID and snapshotID because we are generating new UUID  for each request for the same volume we may generate two volume ID  if restart happens.

#205 fixes the issue as of now. but as suggested by @ShyamsundarR in https://github.com/ceph/ceph-csi/issues/135#issuecomment-461784561 will fix the issue in long run.